### PR TITLE
[WIP] Calculate the runtime complexity of a rule

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
These changes add a new command to OPA which calculates the runtime complexity of the rules in a module. This command can prove useful in determining policy performance and debugging performance issues. The first cut of the implementation does not suppport comprehensions, function calls and the walk builtin.

Signed-off-by: Ashutosh Narkar <anarkar4387@gmail.com>

### CODE REVIEW

File: cmd/analyze.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		os.Exit(analyzeModules(args))
+	},
+}
+
+func analyzeModules(args []string) int {
+	modules := map[string]*ast.Module{}
+
+	f := loaderFilter{}
+
+	result, err := loader.Filtered(args, f.Apply)
```
Comment (patrick-east) [inline cmd/analyze.go]: We should add bundle support, we started pushing all the CLI's to support `-b/--bundle` for loading modules so we should keep it up for new ones. It should be pretty easy to copy what we do in some of the other CLI's.

File: internal/cmd/analyze/analyze.go [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+		resultComplexity = "O(1)"
+	}
+
+	// remove end brackets
+	resultComplexity = strings.TrimPrefix(resultComplexity, "[")
+	return strings.TrimSuffix(resultComplexity, "]")
+}
+
+// runtimeComplexityExpr analyzes an expression and returns its runtime complexity
+func runtimeComplexityExpr(expr *ast.Expr, analyzeExprResult *analyzeExpression, analyzeRuleResult *analyzeResult, all []*analyzeResult) {
```
Comment (patrick-east) [inline internal/cmd/analyze/analyze.go]: nit: Can some of these bigger functions be split up into some helper functions? Can maybe separate the code that does the condition checks for type of expression/term and the handling of the types.

File: internal/cmd/analyze/analyze.go [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+//
+// Example:
+// Rule: allow  Complexity: O(foo)
+// Rule: foo    Complexity: O(input.bar)
+//
+// becomes
+//
+// Rule: allow  Complexity: O(input.bar)
+// Rule: foo    Complexity: O(input.bar)
+func postProcessComplexityResults(preRes map[string][]string) map[string][]string {
```
Comment (tsandall) [inline internal/cmd/analyze/analyze.go]: NOTE: I haven't reviewed the post processing in-depth yet. I'd like to come back to this after addressing the other comments.

**Comment by tsandall** [comment]
Closing this for now. We'll revisit this for v0.21.

**Pull request closed without merging**